thor/main_2.py

```diff
@@ -122,26 +121,123 @@ def link_test_orbit(
     ],
     propagator: Propagator = PYOORB(),
     max_processes: int = 1,
-):
+) -> Iterator[Any]:
     """
-    Find all linkages for a single test orbit.
+    Run THOR for a single test orbit on the given observations. This function will yield
+    results at each stage of the pipeline.
+        1. transformed_detections
+        2. clusters, cluster_members
+        3. iod_orbits, iod_orbit_members
+        4. od_orbits, od_orbit_members
+        5. recovered_orbits, recovered_orbit_members
+
+    Parameters
+    ----------
+    test_orbit : `~thor.orbit.TestOrbit`
+        Test orbit to use to gather and transform observations.
+    observations : `~thor.observations.observations.Observations`
+        Observations from which range and transform the detections.
+    filters : list of `~thor.observations.filters.ObservationFilter`, optional
+        List of filters to apply to the observations before running THOR.
+    propagator : `~adam_core.propagator.propagator.Propagator`
+        Propagator to use to propagate the test orbit and generate
+        ephemerides.
+    max_processes : int, optional
+        Maximum number of processes to use for parallelization.
     """
+    # Apply filters to the observations
+    filtered_observations = observations
+    if filters is not None:
+        for filter_i in filters:
+            filtered_observations = filter_i.apply(filtered_observations, test_orbit)
+
     # Range and transform the observations
     transformed_detections = range_and_transform(
         test_orbit,
-        observations,
-        filters=filters,
+        filtered_observations,
         propagator=propagator,
         max_processes=max_processes,
     )
+    yield transformed_detections
+
+    # Translate transformed detections into format required by the rest of the pipeline
+    # FIXME: Remove this translation step
+    observations_df = filtered_observations.to_dataframe()
+    observations_df.rename(
+        columns={
+            "detections.id": "obs_id",
+            "detections.ra": "RA_deg",
+            "detections.dec": "Dec_deg",
+            "detections.ra_sigma": "RA_sigma_deg",
+            "detections.dec_sigma": "Dec_sigma_deg",
+            "detections.mag": "mag",
+            "detections.mag_sigma": "mag_sigma",
+        },
+        inplace=True,
+    )
+    observations_df["mjd_utc"] = (
+        filtered_observations.detections.time.rescale("utc")
+        .mjd()
+        .to_numpy(zero_copy_only=False)
+    )
+
+    # TODO: I think we might want to add state_id to the observers table. Essential wrap the observers class with another table.
```

Comment:
Yes, something like this will be necessary I think.

Reply:
Added in this PR.